plagiarism/process.go

```diff
@@ -0,0 +1,136 @@
+    SubmissionDir string
+    Mossid uint32
+    Logger *log.Logger
+    ResultDir string
+}
+
+// DecodeRequest decode request from request body
+func DecodeRequest(body io.ReadCloser) ([]string, error) {
+    decoder := json.NewDecoder(body)
+    var guidsStruct GUIDsStruct
+    err := decoder.Decode(&guidsStruct)
+
+    if err != nil {
+        return []string{}, err
+    }
+    return guidsStruct.GUIDs, nil
+}
+
+// GUIDPath use to get submission from OmegaUp
+// Example GUID : abcdefghi1234
+// GUID Path    : submissionDir/ab/cdefghi1234
+func (analysis *Analysis) GUIDPath(guid string) string {
+    return fmt.Sprintf("%v/%v/%v", analysis.SubmissionDir, guid[:2], guid[2:])
+}
+
+// GUIDResultDir use to specify in what directory the analysis results are stored
+// Example GUID : abcdefghi1234
+// GUID Path    : resultDir/ab
+func (analysis *Analysis) GUIDResultDir(guid string) string {
+    return fmt.Sprintf("%v/%v", analysis.ResultDir, guid[:2])
+}
+
+// GUIDResultPath use to specify in what path the analysis results are stored
+// Example GUID : abcdefghi1234
+// GUID Path    : resultDir/abcdefghi1234
+func (analysis *Analysis) GUIDResultPath(guid string) string {
+    return fmt.Sprintf("%v/%v/%v", analysis.ResultDir, guid[:2], guid[2:])
+}
+
+// RunDetector requests MOSS server to run plagiarism detector
+func (analysis *Analysis) RunDetector() Response {
+    m := gomoss.New(analysis.Mossid)
+    response := Response{
+        Status: "ok",
+    }
+
+    //Initialize logger
+    logger := analysis.Logger
+
+    // Add codes from submission path
+    for _, guid := range analysis.GUIDs {
+        path := analysis.GUIDPath(guid)
+        m.AddCodeCustomName(path, guid)
+    }
+
+    // Request analysis to MOSS
+    // TODO (SpaceWhite) error return on gomoss request
+    url := m.Request()
+
+    // Get extracted results from given URL
+    // TODO (SpaceWhite) error return on gomoss extractMoss
+    results := gomoss.ExtractMoss(url)
+
+
+    // Store analysis results into resultdir path
+    for _, guid := range analysis.GUIDs {
+        path := analysis.GUIDResultPath(guid)
+
+        // Create a path when the GUID is not exist
+        os.Mkdir(analysis.GUIDResultDir(guid), os.ModePerm)
+        f, err := os.Create(path)
+
+        // Fail when can't create GUID result file
+        if err != nil {
+            logger.Printf("Can't create %v file: %v\n", path, err)
+            response.Status = "fail"
+            return response
+        }
+
+        similarity := make(map[string][]interface{})
+        for _, result := range results {
+            if result.Codes[0] == guid || result.Codes[1] == guid {
+                similarity[guid] = append(similarity[guid], result)
+            }
+        }
+
+        // Store results into encoded JSON
+        json.NewEncoder(f).Encode(similarity)
+        f.Close()
+    }
+    return response
+}
+
+// GetResults from specified directory
+func (analysis *Analysis) GetResults() Response {
```

Comment:
How about storing the results in a (gzipped) JSON, so that you only need to ungzip it and send it over? IIRC we had agreed about this in the design doc.